Registered Arrays and Graph::insert

N-Coder opened the pull request:
Continuation of PR [\#539 (internal link)](https://git.tcs.uos.de/ogdf-devs/OGDF/-/merge_requests/539) from GitLab. Now includes the `RegisteredArrays` plus the rewritten `Graph::insert` and further fixes. The below change list can probably be directly copied into the release note in `doc/relnotes` once those are created.

# Changes

## RegisteredArrays
- Unified `NodeArray`, `EdgeArray`, `AdjEntryArray`, `FaceArray`, `ClusterArray`, `HypernodeArray`, and `HyperedgeArray` to a common implementation based on `RegisteredArray`.
- Labelling in further "universes" / for further sets of indexable objects can easily be provided by implementing a custom `Registry`.
- `RegisteredArray` uses `std::vector` instead of `ogdf::Array` for value storage. This means that when growing, elements will be copied when their move constructor is not `noexcept` (see the porting notes).
- Setting the second template parameter of `RegisteredArray` to `false` now allows it to accept non-copy-initializable classes like `std::unique_ptr<Graph>` as values. `NodeArrayP<Graph>` (or similar for all the above types) provides a short-cut for this usage to, e.g. in this case, label each of a Graph's nodes with another Graph.
- Added common implementation of `RegisteredSet` and added accompanying instantiations for each of the above types.  Removed `ClusterSetSimple` in favour of `ClusterArray<bool>` and `ClusterSetPure` in favour of `ClusterSet<false>`.
- Ensured `RegisteredSet::clear` has runtime linear in its number of elements.

## GraphCopy, Graph::insert and GraphObserver
- `GraphCopySimple` and `GraphCopy` now share `GraphCopyBase` as parent class.
- `GraphCopyBase::setOriginalEmbedding` now also works after the GraphCopy was modified (on a best-effort basis).
- Merge `Graph(Copy(Simple))::construct`, `::assign`, `::copy`, `::(construct)InitByXYZ` and `ogdf::inducedSubgraph` to single `Graph::insert` implementation. This implementation is also used when copy-constructing or (in combination with clear) when copy-assigning. It replaces the following different implementations:
  ```c++
  void Graph::construct(const Graph &G, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void Graph::assign(const Graph &G, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void Graph::insert(const Graph &G);
  void Graph::insert(const Graph &G, NodeArray<node> &nodeMap);
  void Graph::copy(const Graph &G);
  void Graph::copy(const Graph &G, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void Graph::constructInitByNodes(const Graph &G, const List<node> &nodeList, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void Graph::constructInitByActiveNodes(const List<node> &nodeList, const NodeArray<bool> &activeNodes, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void Graph::constructInitByCC(const CCsInfo &info, int cc, NodeArray<node> &mapNode, EdgeArray<edge> &mapEdge);
  void GraphCopySimple::init(const Graph &G);
  void GraphCopySimple::initGC(const GraphCopySimple &GC, NodeArray<node> &vCopy, EdgeArray<edge> &eCopy);
  void GraphCopy::init(const Graph &G);
  void GraphCopy::initByCC(const CCsInfo &info, int cc, EdgeArray<edge> &eCopy);
  void GraphCopy::initByNodes(const List<node> &origNodes, EdgeArray<edge> &eCopy);
  void GraphCopy::initByActiveNodes(const List<node> &nodeList, const NodeArray<bool> &activeNodes, EdgeArray<edge> &eCopy);
  void GraphCopy::initGC(const GraphCopy &GC, NodeArray<node> &vCopy, EdgeArray<edge> &eCopy);
  // from <extended_graph_alg.h>:
  void inducedSubGraph(const Graph &G, LISTITERATOR start, Graph &subGraph);
  void inducedSubGraph(const Graph &G, LISTITERATOR start, Graph &subGraph, NodeArray<node> &nodeTableOrig2New);
  void inducedSubGraph(const Graph &G, LISTITERATOR start, Graph &subGraph, NodeArray<node> &nodeTableOrig2New, EdgeArray<edge> &edgeTableOrig2New);
  void inducedSubGraph(const Graph &G, LISTITERATOR start, GraphCopySimple &subGraph);
  ```
- `GraphCopyBase::insert` will automatically update its mappings when inserting from the original Graph
- Refactored `Graph::newEdge` and `Graph::newNode` to simpler implementations with less duplications and also support further parameter combinations of `newEdge`.
- Notify `GraphObserver`s when their `Graph` is destructed through `GraphObserver::registrationChanged()`.
- Add common base classes `Observable` + `Observer` for this pattern.

## Various
- When compiling with C++20, added concepts OGDF_(NODE|EDGE)_(FILTER|ITERATOR|LIST) that match types and make working with templates (or understanding template errors) much easier. Macro `OGDF_HAS_CONCEPTS` is only defined if these concepts are available. For earliert C++ versions, the names are macros resolving to `typename` for easier usage in `template<OGDF_NODE_LIST ListType>`. 
- Macro `OGDF_CHECK_CONCEPT` can be used to check statically concepts when they are available, e.g. in `OGDF_CHECK_CONCEPT(OGDF_NODE_ITER<internal::GraphIterator<node>>)`. 
- Many ogdf iterators are now also valid std iterators down to in C++11 and are statically checked in C++20.
- `GraphIteratorBase` now yields `const` pointers and has a const version.
- Added `OGDFAllocator` as allocator for `std` containers that uses the OGDF memory manager.
- Marked multiple constructors as `noexcept`.
- Added `OGDF_DISABLE_WARNING_PUSH`, `OGDF_DISABLE_WARNING_THROW_TERMINATE`, `OGDF_DISABLE_WARNING_POP`.
- Clean-up in `orthogonal/edge_router/NodeInfo.h`.
- Cleaned up some internals `ClusterGraph`.
- Add attribute `OGDF_NODISCARD` for the corresponding C++ attribute.
- `CCsInfo` now provides range-like iterators for the nodes and edges of each components.

N-Coder commented on include/ogdf/basic/GraphCopy.h:
```diff
@@ -73,8 +73,6 @@ class OGDF_EXPORT GraphCopySimple : public Graph {
 	//! Copy constructor.
 	GraphCopySimple(const GraphCopySimple& GC);
 
-	virtual ~GraphCopySimple() { }
```
ToDo: where does this come from?

N-Coder replied:
derived from Graph, so there's already a virtual destructor

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -526,12 +702,10 @@ class OGDF_EXPORT Graph {
+	GraphRegistry<NodeElement> m_regNodeArrays; //!< The registered node arrays.
+	GraphRegistry<EdgeElement> m_regEdgeArrays; //!< The registered edge arrays.
+	GraphRegistry<AdjElement, GraphAdjIterator> m_regAdjArrays;
+	GraphAdjIterator m_adjIt;
```
ToDo: maybe this can be converted to some `begin(obj), end(obj)` magic?

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -634,13 +808,33 @@ class OGDF_EXPORT Graph {
+	int adjEntryArrayTableSize() const { return m_regEdgeArrays.getArraySize() << 1; }
+
+	//! Returns a reference to the registry of node arrays associated with this graph.
+	GraphRegistry<NodeElement>& nodeRegistry() { return m_regNodeArrays; }
```
ToDo: do we need a non-const version here?

N-Coder commented on include/ogdf/cluster/ClusterGraphAttributes.h:
```diff
@@ -110,8 +110,6 @@ class OGDF_EXPORT ClusterGraphAttributes : public GraphAttributes {
 	explicit ClusterGraphAttributes(const ClusterGraph& cg,
 			long initAttributes = nodeGraphics | edgeGraphics | clusterGraphics);
 
-	virtual ~ClusterGraphAttributes() { }
```
?

N-Coder replied:
already made virtual by parent class

N-Coder commented on include/ogdf/cluster/ClusterSet.h:
```diff
@@ -211,89 +75,21 @@ class OGDF_EXPORT ClusterSetPure {
+	const List<cluster>& clusters() const { return RS::elements(); }
 };
 
+using ClusterSetSimple = ClusterSetPure;
```
Do we need this for something?

N-Coder commented on include/ogdf/orthogonal/edge_router/NodeInfo.h:
```diff
@@ -79,53 +83,6 @@ class OGDF_EXPORT NodeInfo {
-
-		for (int i = 0; i < 4; ++i) {
-			// these should work with move cstrs once VC++ implements rvalue references v3
-			in_edges[i] = std::move(other.in_edges[i]);
```
todo check that this works

N-Coder commented on src/ogdf/cluster/ClusterGraph.cpp:
```diff
@@ -249,7 +249,7 @@ void ClusterGraph::shallowCopy(const ClusterGraph& C) {
 // Initialize the graph
 void ClusterGraph::initGraph(const Graph& G) {
-	reregister(&G); //will in some constructors cause double registration
+	reregister(&G); //FIXME will in some constructors cause double registration
```
FIXME

N-Coder commented on src/ogdf/basic/Graph.cpp:
```diff
@@ -1038,32 +962,25 @@ void Graph::unregisterStructure(ListIterator<GraphObserver*> it) const {
-	for (NodeArrayBase* nab : m_regNodeArrays) {
-		nab->reinit(m_nodeArrayTableSize);
-	}
+void Graph::reinitArrays() {
```
ToDo check all .cpp implementations (also for ClusterGraph)

N-Coder commented on include/ogdf/basic/RegisteredArray.h:
```diff
@@ -0,0 +1,843 @@
+		}
+		m_size = size = max(size, 0);
+		for (registered_array_type* ab : m_registeredArrays) {
+			ab->resize(size, shrink);
```
do we want to keep old entries around?

N-Coder commented:
Ok, these changes are slowly taking shape. The commits still need some rebasing and squashing and we still need some good tests (except for all the implicit ones through usages of `GraphCopy`), but then this PR (which might need switching to the [nfink-subgraph](https://github.com/ogdf/ogdf/tree/nfink-subgraph) branch as its hard to entirely decouple the `RegisteredArray` and `insert` changes) would be good to go. There are still two important API design decision to be made where I would need some feedback (:eyes: @milsen).

First, how should we name our Arrays. Currently, they look like this:
```cpp
template<typename Value>
using NodeArray = GraphRegisteredArray<NodeElement, Value, true>;

template<typename Value>
using NodeArrayWithoutDefault = GraphRegisteredArray<NodeElement, Value, false>;

template<typename Value>
using NodeArrayOfPtrs = GraphRegisteredArray<NodeElement, std::unique_ptr<Value>, false>;
```
where the boolean template parameter is `WithDefault` indicating whether we should only default construct the contained values (`false`) or should store an explicit default value and then copy construct the contained values from this (`true`, which won't work for classes with deleted copy constructors like `unique_ptr`). As having an Array of Pointers that need deletion afterwards is a somewhat common use-case, I provided the `NodeArrayOfPtrs` as a short-cut for this. Effectively, the semantics of `NodeArray` stay the same, the two other variants are new, although I'm not super happy with their names. Any better suggestions?

Second, I managed to reduce the diversity of insert variants down to three:
1. Insert from a list (i.e. iterator) of nodes and an optional filter (e.g. `std::function<bool(edge)>`) of edges, which might always return true.
2. Insert from a list of nodes and a list of edges, where we assume the edges to be some already stored subset of all edges incident to the given nodes.
3. Syntatic sugar methods, like inserting a whole Graph or a whole connected component from CCsInfo, which basically reduces to 1. with a list of nodes and all incident edges, i.e. a filter that always returns true.
```cpp
template<OGDF_NODE_ITER NI, OGDF_EDGE_FILTER EF, bool copyEmbedding = true,
		bool copyIDs = false, bool notifyObservers = true>
std::pair<int, int> insert(const NI& nodesBegin, const NI& nodesEnd, const EF& edgeFilter,
		NodeArray<node>& nodeMap, EdgeArray<edge>& edgeMap); // 1

template<OGDF_NODE_ITER NI, OGDF_EDGE_ITER EI, bool copyEmbedding = true, bool copyIDs = false,
		bool notifyObservers = true>
std::pair<int, int> insert(const NI& nodesBegin, const NI& nodesEnd, const EI& edgesBegin,
		const EI& edgesEnd, NodeArray<node>& nodeMap, EdgeArray<edge>& edgeMap); // 2

std::pair<int, int> insert(const Graph& G, NodeArray<node>& nodeMap, EdgeArray<edge>& edgeMap); // 3
```
The main difference is that in 1. we can use the filter to iterate through all adjEntries and thereby easily copy the embedding on the fly. If we are given a list of edges as in 2., this is not possible and we need to first iterate through this list to copy the edges and then through all adjEntries to copy the embedding. Thus, the behaviour of the methods for 1 and 2 is slightly different:

**1. `insert(NI, EI)`**
- passes:
  - one pass over nodes <- node creation
  - early exit if no edges
  - one pass over edges <- edge creation
  - early exit if !copyEmbedding
  - one pass over nodes+adjEntries <- embedding
  - early exit if !notifyObservers
  - one pass over edges <- notifications
- nodeMap & edgeMap overwritten if non-null, edge created when both endpoints in NI or in nodeMap
- !copyEmbedding: embedding follows order of G.edges

**2. `insert(NI, EF)`**
- passes:
  - one pass over nodes <- node creation
  - one pass over nodes+adjEntries <- edge creation + embedding
  - early exit if !notifyObservers
  - one pass over nodes+adjEntries <- notifications (for source adj of all edgeMap entries)
- nodeMap overwritten if non-null, edgeMap not overwritten, edge invalid when one old endpoint in nodeMap
- embedding always copied

The main question here is whether we want to have different names for these two methods and thus let the user decide explicitly which behaviour they want, or whether we want to let this happen automatically, possibly making it more intransparent to the user what actually happens.


Also, there are some short-cuts which we might want to define, but which also exponentially increase the number of overloaded methods:
- pass lists/vectors instead of `begin()` + `end()`
- do not pass a `nodeMap` or `edgeMap`
- pass a node filter instead of a node iterator (then also requires passing the underlying graph for filtering its nodes)
Which of these overloads do we want to create? My suggestion would be only one, namely `insert(const Graph&)`, as this is very common. For all other points, the user can easily wrap their data to have the correct format and this keeps our API cleaner and simpler.

milsen commented:
Answer to https://github.com/ogdf/ogdf/pull/101#issuecomment-1790973591: Ok, I talked with Markus about this (not sure whether that was a good idea :wink:). I will just use the names you initially suggested, so it's clear what I'm referring to.

## Arrays
### `NodeArrayWithoutDefault`:
In general, we should not use the term "default" because it is not clear at all whether we mean the default value given by a parameter or the default value given by C++ in the default constructor (which would be the opposite of what we mean here).

A question: Why is it not possible to have the `NodeArrayWithoutDefault` functionality as part of NodeArray? Basically a compile-time check inside NodeArray: If copy-constructable, then NodeArrayWithDefault, else NodeArrayWithoutDefault?
If there are good reasons, then a follow-up question: Do you have concrete use cases for `NodeArrayWithoutDefault` that are not covered by `NodeArrayOfPtrs`?
If not, we may not need the new type. If there are, however, some use cases, then the new type maybe should be allowed to have a longer, clunkier name - such that it isn't accidentally used by people that actually just want a normal NodeArray and don't know what they are doing. Maybe `NodeArrayOfUniques` (hm, could be confused with `unique_ptr`... :confused:) or `NodeArrayOfNonCopyables` is ok?

### `NodeArrayOfPtrs`
Markus does not like that it sounds like the array contains c-style pointers (i.e. `*myptr`) rather than `unique_ptr`s. 
He suggests `NodeArrayP`. While it is not very descriptive, I do see his point: It's short because it's used often and yet, people will have to read the doc to understand, which might be good in this case actually. Also "p_myThing" is often used for pointers so the P also has that association with it. I'm not really happy with it, though... :thinking:

### Mirkos Solution
@wirkom suggests to only have one NodeArray to rule them all and use a template param for unique_ptrs.
```
template<typename Value, bool useUniquePtrs = false>
using NodeArray = typename std::conditional<useUniquePtrs,
	GraphRegisteredArray<NodeElement, std::unique_ptr<Value>, false>,
	GraphRegisteredArray<NodeElement, Value, true>>::type;
```

## insert
Markus and I agreed on:
* :heavy_check_mark: insert for containers (list, vector) without begin, end (will be used often)
* :x: insert without nodeMap, edgeMap (you need those params in most cases)
* :x: insert with Graph + NodeFilter (if you have a filter, just build your own node list)
* :heavy_check_mark: insert(const Graph&) (will be used often)
* :heavy_check_mark: insert(const CCsInfo&) (will be used often)

N-Coder commented:
Thanks for your input! There already is a boolean flag for selecting the copy-constructing or the default-constructing array variant, I just thought that differentiating this via classname instead of a boolean parameter might be easier for users. I now changed the Array definition to also expose this boolean parameter, but now with a default value that selects the copy mode only if the passed class is copy constructible. Note that this means that the interface of the class `NodeArray<Value>` (especially the parameters its constructor takes) will be slightly different depending on the contained `Value`. I dropped the `WithoutDefault` variant and renamed the `unique_ptr` variant (as selecting this with a further boolean flag would not be much shorter but possibly more confusing than explicitly passing `unique_ptr<Value>`).
```cpp
template<typename Value, bool copy_default = std::is_copy_constructible<Value>::value>
using NodeArray = GraphRegisteredArray<NodeElement, Value, copy_default>;
template<typename Value>
using NodeArrayP = NodeArray<std::unique_ptr<Value>>;
 ```

I also changed the `insert` implementations accordingly. Now I just need to write a few good test, then this PR should be ready for review. The remaining points from the above todo list are ideas for further improvements that can be done separately from this PR.

milsen reviewed (changes requested):
Ok, I'm done with the review! First of all, thank you so much for the work; it is very impressive and solves a lot of issues with the current implementation. :butterfly:

I have some remarks, often times just things I may not have understood correctly because I lost the overview over everything at some point :sweat_smile:
There are also many additional remarks I have not commented on for now, because they only concern typos and clarity issues etc. in the documentation. I think it is easier if I just make those changes myself directly after you have made changes for my current review. So my plan for how to continue would be this:
1. You respond to my review/push changes.
2. Afterwards, I make my changes to the comments/docs, and then squash some commits (e.g. the commits for adding RegisteredArray, Observer, and changing Graph_d.h – those did not make a lot of sense when separated because e.g. some methods were defined first in one commit and then declared in a later commit).
3. Force-push + merge.
4. Party.

N-Coder commented on include/ogdf/basic/internal/config.h:
```diff
@@ -130,6 +130,80 @@ using std::to_string;
+
+//! Disable the warning that some computed value is not used
+//! @ingroup macros
+#define OGDF_DISABLE_WARNING_UNUSED_VALUE
```
Removed

N-Coder commented on include/ogdf/basic/GraphList.h:
```diff
@@ -292,6 +295,15 @@ class OGDF_EXPORT GraphListBase {
+	//! Provides a bidirectional const iterator to an object in the container.
+	using const_iterator = GraphConstIterator<T*>;
+	//! Provides a bidirectional reverse iterator to an object in the container.
+	using reverse_iterator = GraphReverseIterator<T*>;
```
See [here](https://github.com/ogdf/ogdf/pull/101#discussion_r1463328422)

N-Coder commented on include/ogdf/basic/internal/graph_iterators.h:
```diff
@@ -40,61 +40,67 @@ namespace ogdf {
-using GraphReverseIterator = GraphIteratorBase<GraphObjectPtr, true>;
+using GraphConstIterator = GraphIteratorBase<GraphObjectPtr, false, true>;
+template<class GraphObjectPtr>
+using GraphReverseIterator = GraphIteratorBase<GraphObjectPtr, true, false>;
```
I tried introducing the whole const iterator concept to graph lists while making them behave like standard iterators following the same way of thinking. 😅  This then broke the non-const method `const GraphObjectPtr& GraphIteratorBase::operator*()` for some weird reason I no longer recall (see the change disabling it [here](https://github.com/ogdf/ogdf/commit/c38e38c9422625bd378c402ce6c2fdb5718bf8b8#diff-72053da2e3249e43a9905b7d3c742e8dd74c4991b16206da838616bc4ba6b932), unfortunately I no longer find the failing CI run before that) - probably the method was already broken as it was, the change to const iterators just surfaced this by breaking compilation. Constness probably makes little sense anyways with GraphList iterators, because neither the list nor the contained objects allows public modification. I guess this whole tangle of classes used for graph element storage (esp. the empty `GraphObjectContainer` class or the list destructors!) needs a further clean-up, but that should be done in another PR. The only usage of these `using`s is in GraphList.h anyways (and the std-lib now introduced its own wrappers for reversing and const iterators), so they could probably also go as a whole...

N-Coder commented on include/ogdf/basic/RegisteredArray.h:
```diff
@@ -0,0 +1,921 @@
+private:
+	mutable registration_list_type m_registeredArrays;
+	bool m_autoShrink = false;
+	int m_size = 0;
```
I put evaluating this change on the follow-up todo list in #225 as I assume this can be done independently of this PR

N-Coder commented on include/ogdf/basic/RegisteredArray.h:
```diff
@@ -0,0 +1,921 @@
+protected:
+	//! Returns a reference to the associated registry.
+	inline const Registry& getRegistry() const {
+		OGDF_ASSERT(registeredAt());
```
Yes, but this makes it explicit in the code that we check for nullptr-dereferences and doesn't really cost us anything. I guess the `valid` check is more ephemeral as it just adds an occasional check that no prior resizing was missed.

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -456,26 +514,273 @@ void NodeElement::outEdges(EDGELIST& edgeList) const {
+
+public:
+	using iterator = GraphAdjIterator;
+	using const_iterator = GraphAdjIterator;
```
I guess this is the same const-iterators issue as [above](https://github.com/ogdf/ogdf/pull/101#discussion_r1466084611), where all GraphObject containers plus their iterators need a further clean-up esp. with regards to how to handle const-correctness. For now, `const_iterator` is not used and I thus removed it.

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -703,57 +993,35 @@ class OGDF_EXPORT Graph {
 	 * @return the newly created node.
 	 */
-	node newNode(int index);
+	node newNode(int index = -1) {
```
Is your comment referring to the overloaded `newEdge` as there is only one `newNode` method? For both, the `inline`s are currently set such that (without the compiler deciding for further inlining) each call site of the often-used `newNode`/`newEdge` methods contains exactly one function call, which directly refers to the function doing the actual creation without further jumps. The only exception to this is the not-necessarily inlined but private `pureNewEdge` called by the templated ("master") `newEdge`, as `pureNewEdge` is not super short and used multiple times, so the compiler should decide itself about the inlining.

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -1352,6 +1628,227 @@ class OGDF_EXPORT Graph {
+	 * connected component \p cc as described by CCsInfo \p info.
+	 */
+	std::pair<int, int> insert(const CCsInfo& info, int cc, NodeArray<node>& nodeMap,
+			EdgeArray<edge>& edgeMap) {
```
Added!

N-Coder commented on include/ogdf/basic/Graph_d.h:
```diff
@@ -1516,3 +2067,5 @@ inline std::ostream& operator<<(std::ostream& os, const NodePair& np) {
 
 }
+
+#include <ogdf/basic/InducedSubgraph.h>
```
I was thinking about cleaning up the huge `Graph_d.h` header (consequentially removing it in favor of `Graph.d`) and splitting out its parts into files in a internal subdirectory which would then also include `InducedSubgraph.h`. This header not being inlined is already in anticipation of this, while I'd do that move only in the later PR.

N-Coder commented on include/ogdf/basic/InducedSubgraph.h:
```diff
@@ -0,0 +1,298 @@
+		node vG = *it;
+		node v = nodeMap[vG];
+		for (adjEntry adjG : vG->adjEntries) {
+			edge eG = adjG->m_edge;
```
Done

N-Coder commented on include/ogdf/basic/InducedSubgraph.h:
```diff
@@ -0,0 +1,298 @@
+			// edgeMap[eG] might be an old entry that was already inserted into the list
+			// so check whether adj is already in the list, indicated by having succ or pred,
+			// or being the only entry in the list
+			if (adj->succ() != nullptr || adj->pred() != nullptr || v->adjEntries.head() == adj) {
```
Yes, the idea was to allow re-use of nodeMap and edgeMap, not only for multiple CCs or potentially even arbitrary subgraphs, but also for `GraphCopy.insert(G)` to directly pass its own maps instead of creating local ones on the stack. Unfortunately, this is not so easy for the filtering version, which is why the docs currently forbid this and `GraphCopy.insert` still uses additional maps on the stack. It would be nice if we could allow this in a well-defined manner, but that probably requires more work (esp. with different possible states in GraphCopy) and I'm for now not sure whether that is worth it for this niche use-case.

N-Coder commented on include/ogdf/basic/GraphCopy.h:
```diff
@@ -180,6 +191,166 @@ class OGDF_EXPORT GraphCopySimple : public Graph {
+	 */
+	virtual void setOriginalEmbedding() = 0;
+
+	bool isLinkCopiesOnInsert() const { return m_linkCopiesOnInsert; }
```
Renamed

N-Coder commented on include/ogdf/cluster/ClusterGraph.h:
```diff
@@ -294,9 +298,37 @@ inline bool test_forall_adj_edges_of_cluster(adjEntry& adj, edge& e) {
+	virtual void clusterDeleted(cluster v) = 0;
+	virtual void clusterAdded(cluster v) = 0;
+
+	// virtual void cleared() = 0;
```
Fixed, added `ClusterGraphObserver::clustersCleared()`

N-Coder commented on src/ogdf/basic/GraphCopy.cpp:
```diff
@@ -112,202 +91,150 @@ void GraphCopySimple::delEdge(edge e) {
-				m_eIterator[eCopy[e]] = m_eCopy[e].pushBack(eCopy[e]);
-				m_eOrig[eCopy[e]] = e;
+			edge e = copy(adj->theEdge());
+			if (e != nullptr && e->isIncident(u)) {
```
I remember initially using the same `setOriginalEmbedding` code for all versions of `GraphCopyBase`, but this broke in `GraphCopy`. The issue was probably that `copy(adj)` sometimes returns adjEntries with `copy(adj)->theNode() != copy(adj->theNode())` (maybe because it per its docs doesn't care about reversed edges?). My simple solution to this was splitting the method into [two versions](
https://github.com/ogdf/ogdf/commit/60f9a8ecc893c3443e9136327759ca0301315534#diff-90c54bd7a34123d3157346d20b16a48e9582030899d09a70beb49d4bd1014343) and simply treating such edges as new/unembedded ones whenever this happens for `GraphCopy`. We thereby still support strictly more use-cases than before, but this could probably be further improved. Maybe checking for reversed edges is already enough, but I'd leave that to someone with more experience with the workings of the non-simple `GraphCopy`.

N-Coder commented on include/ogdf/basic/GraphCopy.h:
```diff
@@ -38,51 +38,66 @@
 
-	virtual ~GraphCopySimple() { }
+	//! Re-initializes the copy using \p G, but does not create any nodes or edges.
+	// OGDF_DEPRECATED("use setOriginalGraph")
```
I removed all the usages. Should we leave the deprecated methods in or remove them entirely?

milsen replied:
I'll just keep them removed :man_shrugging:

N-Coder commented on test/src/basic/adjentryarray.cpp:
```diff
@@ -55,8 +55,69 @@ go_bandit([]() {
-			{1, 2, 3}, {42}, chooseAdjEntry, allAdjEntries, createAdjEntry);
+	auto init = [](Graph& graph) { randomGraph(graph, 42, 168); };
+
+	describeArray<Graph, AdjEntryArray, adjEntry, int>( //
```
Feel free to refactor the tests however you like! ;)

N-Coder commented on test/src/basic/array_helper.h:
```diff
@@ -35,189 +35,503 @@
+				array->init();
+				AssertThat(array->registeredAt(), IsNull());
+				AssertThat(array->valid(), IsFalse());
+			});
```
I tried doing this with templates, but there are some slight differences with how default values are created and handled, which made my two attempts at doing this fail. It might now be easier with C++17 `constexpr if`, but I settled for being ok with this slight duplication (although Sonar also still complains about this).

N-Coder commented on test/src/basic/graphcopy.cpp:
```diff
@@ -220,6 +217,86 @@ void describeGraphCopySimple(int numberOfNodes) {
+			graphCopy.reset(new GCType());
+			int numberOfCC = ccs.numberOfCCs() - 1;
+			graphCopy->createEmpty(graph);
+			graphCopy->initByCC(ccs, numberOfCC, eCopy);
```
Resolved

N-Coder commented on test/src/basic/graphcopy.cpp:
```diff
@@ -220,6 +217,86 @@ void describeGraphCopySimple(int numberOfNodes) {
+			List<node> asdf;
+			graphCopy->allNodes(asdf);
+			List<edge> asdfgh;
+			graphCopy->allEdges(asdfgh);
```
They were not even used, so removed them

N-Coder commented on test/src/basic/graphobserver.cpp:
```diff
@@ -0,0 +1,406 @@
+			edge e = G1.newEdge(G1.newNode(), G1.newNode());
+			AssertThat(e->index(), Equals(0));
+			edge f = G1.split(e);
+			AssertThat(f->index(), Equals(1));
```
And changing how indices work would probably also affect users' code that implicitly relies on their behaviour even if that is not explicitly documented. Users will consider any publicly observable behaviour as part of the API anyways, so it might be even nice to have these tests fail if we change something about that behaviour.

N-Coder commented on test/src/basic/graphobserver.cpp:
```diff
@@ -0,0 +1,406 @@
+				AssertThat(GO.m_events,
+						Equals(std::vector<std::pair<GOEvent, int>> {{GOEvent::NODE_ADD, 2}}));
+			}
+			G1.newNode(); // a callback now would cause a memory error
```
None of the added Observers are in scope anymore, so calling methods on them (e.g. if they were not cleanly unregistered once they went out of scope) would make ASAN complain, indirectly causing this test to fail.

N-Coder commented on test/src/basic/insert.cpp:
```diff
@@ -0,0 +1,506 @@
+	ogdf::makeConnected(G);
+	ogdf::planarEmbed(G);
+
+	CombinatorialEmbedding emb(G);
```
I'm not 100% sure why I re-implemented `triangulate` as I would have used that function if it would have fit. It might be that this function generates a random triangulation while the `simple_graph_alg` function generates very regular ones (many subdivision edges to the same vertex)?

N-Coder commented on test/src/basic/insert.cpp:
```diff
@@ -0,0 +1,506 @@
+	};
+}
+
+#define S(b) std::string(b ? "true" : "false")
```
Could you take care of this?

N-Coder commented on test/src/basic/graphobserver.cpp:
```diff
@@ -0,0 +1,406 @@
+			AssertThat(GO.m_nodesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+
+			AssertThat(GO.m_events.size(), Equals(1 + 1 + 5 + 7 + 1));
```
I tend to ignore these warnings, could you fix this?

N-Coder commented on test/src/basic/graphobserver.cpp:
```diff
@@ -0,0 +1,406 @@
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+			AssertThat(GO.m_edgesDeleted, Equals(0));
+
+			AssertThat(GO.m_events.size(), Equals(1 + 5 + 7));
```
I tend to ignore these warnings, could you fix this?

N-Coder commented:
@milsen feel free to proceed with step 2 ;)
And thanks a lot for taking time for reviewing and improving this PR!

(the MacOS CI failure seems to be due to binaries with incompatible instructions being cached, will investigate)
